machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,131 @@
+	// +kubebuilder:validation:Required
+	DesiredConfig string `json:"desiredConfig"`
+}
+
+// MachineOSBuildStatus describes the state of a build and other helpful information.
+type MachineOSBuildStatus struct {
+	// buildStart describes when this build began
+	// +kubebuilder:validation:Required
+	BuildStartTime metav1.Time `json:"buildStart"`
+	// buildEnd describes when the build ended
+	// +optional
+	BuildEndTime metav1.Time `json:"buildEnd"`
+	// conditions are state related conditions for the build. Valid types are:
+	// BuildPrepared, Building, BuildFailed, BuildInterrupted, BuildRestarted, and Built
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// relatedBuilds contains previous iterations of failed or interrupted builds related to this one.
+	// This will keep track of failed or interrupted build names
+	// I am thinking we keep track of their name and why they failed
+	// +listType=set
+	// +optional
+	RelatedBuilds []PriorMachineOSBuilds `json:"relatedBuilds,omitempty"`
+}
+
+// BuildProgess highlights some of the key phases of a build to be tracked in Conditions.
+type BuildProgress string
+
+const (
+	MachineOSBuildPrepared    BuildProgress = "BuildPrepared"
+	MachineOSBuiling          BuildProgress = "Building"
+	MachineOSBuildFailed      BuildProgress = "BuildFailed"
+	MachineOSBuildInterrupted BuildProgress = "BuildInterrupted"
+	MachineOSBuildRestarted   BuildProgress = "BuildRestarted"
+	MachineOSBuilt            BuildProgress = "Built"
+)
+
+// PriorMachineOSBuilds contains information about related builds
+type PriorMachineOSBuilds struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// buildFailure contains an optional message of why this build ended prematurely.
+	// +optional
+	BuildFailure string `json:"buildFailure"`
+}
+
+// MachineOSObjectReference is used to track various object types related to a build.
+// We use this instead of a full fledged object refernece because we do not need all of the fields.
+type MachineOSObjectReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
+	// osImageURL is the base OSImage we use to build our custom image.
+	// +kubebuilder:validation:Required
+	OSImageURL string `json:"osImageURL"`
+	// finalImagePushSecret is the secret used to connect to a user registry.
+	// +kubebuilder:validation:Required
+	FinalImagePushSecret string `json:"finalImagePushSecret"`
+	// finalImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder:validation:Required
+	FinalImagePullSecret string `json:"finalImagePullSecret"`
```

Comment:
question: shall the secret be named as `BaseImagePullSecret` if it is the secret used to pull the base image?

Reply:
There should be, at least for now. I feel like we might be conflating `MachineOSBuild` and `MachineOSBuildConfig` (although the latter is outside the scope of this PR).

Reply:
I think if we are making these objects exist for the lifetime of a pool they are associated with, it makes sense to fold most (if not all) config and settings into the MachineOSBuild object @cheesesashimi  wdyt?